compiler+runtime/src/cpp/jank/runtime/obj/ratio.cpp

```diff
@@ -0,0 +1,556 @@
+  {
+    auto ratio_data = make_box<obj::ratio_data>(numerator, denominator);
+    if(ratio_data->denominator == 1)
+    {
+      return make_box<obj::integer>(ratio_data->numerator);
+    }
+    auto r(make_box<obj::ratio>());
+    r->data.numerator = ratio_data->numerator;
+    r->data.denominator = ratio_data->denominator;
+    return r;
+  }
+
+  native_real obj::ratio_data::to_real() const
+  {
+    return static_cast<native_real>(numerator) / denominator;
+  }
+
+  native_real obj::ratio::to_real() const
+  {
+    return data.to_real();
+  }
+
+  native_integer obj::ratio::to_integer() const
+  {
+    return data.numerator / data.denominator;
+  }
+
+  void obj::ratio_data::simplify()
+  {
+    int gcd = std::gcd(numerator, denominator);
+    numerator /= gcd;
+    denominator /= gcd;
+
+    if(denominator < 0)
+    {
+      numerator = -1 * numerator;
+      denominator = -1 * denominator;
+    }
+  }
+
+  void obj::ratio::to_string(fmt::memory_buffer &buff) const
+  {
+    format_to(std::back_inserter(buff), FMT_COMPILE("{}/{}"), data.numerator, data.denominator);
+  }
+
+  native_persistent_string obj::ratio::to_string() const
+  {
+    fmt::memory_buffer buff;
+    to_string(buff);
+    return native_persistent_string{ buff.data(), buff.size() };
+  }
+
+  native_persistent_string obj::ratio::to_code_string() const
+  {
+    return to_string();
+  }
+
+  native_hash obj::ratio::to_hash() const
+  {
+    return hash::combine(hash::integer(data.numerator), hash::integer(data.denominator));
+  }
+
+  native_bool obj::ratio::equal(object const &o) const
+  {
+    if(o.type == object_type::integer)
+    {
+      auto const i(expect_object<obj::integer>(&o));
+      return this->data == i->data;
+    }
+
+    if(o.type == object_type::real)
+    {
+      auto const i(expect_object<obj::integer>(&o));
+      return this->data == i->data;
+    }
+
+    if(o.type == object_type::ratio)
+    {
+      return this == expect_object<obj::ratio>(&o).data;
+    }
+
+    return false;
+  }
+
+  native_integer obj::ratio::compare(object const &o) const
+  {
+    return visit_number_like(
+      [this](auto const typed_o) -> native_integer {
+        return (data > typed_o->data) - (data < typed_o->data);
+      },
+      [&]() -> native_integer {
+        throw std::runtime_error{ fmt::format("not comparable: {}", runtime::to_string(&o)) };
+      },
+      &o);
+  }
+
+  native_integer obj::ratio::compare(static_object const &o) const
+  {
+    return (data > o.data) - (data < o.data);
+  }
+
+  object_ptr operator+(obj::ratio_data const l, obj::ratio_data const r)
+  {
+    native_integer denom = l.denominator * r.denominator;
+    native_integer num = l.numerator * r.denominator + r.numerator * l.denominator;
+    return obj::ratio::create(num, denom);
+  }
+
+  obj::ratio_ptr operator+(obj::integer_ptr l, obj::ratio_data const r)
+  {
+    return expect_object<obj::ratio>(obj::ratio_data(l->data, 1LL) + r);
```

Comment:
Is `expect_object` needed here? The `ratio_data + integer_ptr` overload should return a `ratio_ptr`, right?

Reply:
yeah, i was trying to reuse `ratio_data + ratio_data` (which would return object_ptr). but i suppose we can be more efficient.